man/cancel.1
.\"
.\" cancel man page for CUPS.
.\"
.\" Copyright © 2020-2024 by OpenPrinting.
.\" Copyright © 2007-2019 by Apple Inc.
.\" Copyright © 1997-2006 by Easy Software Products.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH cancel 1 "CUPS" "2026-03-06" "OpenPrinting"
.SH NAME
cancel \- cancel jobs
.SH SYNOPSIS
.B cancel
[
.B \-h
.I hostname[:port]
] [
.B \-E
] [
.B \-U
.I username
] [
.B \-a
] [
.B \-u
.I username
] [
.B \-x
] [
.I id
] [
.I destination
] [
.I destination\-id
]
.SH DESCRIPTION
The \fBcancel\fR command cancels print jobs.
If no \fIdestination\fR or \fIid\fR is specified, the currently printing job on the default destination is canceled.
.SH OPTIONS
The following options are recognized by \fBcancel\fR:
.TP 5
.B \-a
Cancel all jobs on the named destination, or all jobs on all
destinations if none is provided.
.TP 5
.B \-E
Forces encryption when connecting to the server.
.TP 5
\fB\-h \fIhostname\fR[\fI:port\fR]
Specifies an alternate server.
Note: This option must occur before all others.
.TP 5
\fB\-U \fIusername\fR
Specifies the username to use when connecting to the server.
.TP 5
\fB\-u \fIusername\fR
Cancels jobs owned by \fIusername\fR.
.TP 5
.B \-x
Deletes job data files in addition to canceling.
.SH CONFORMING TO
Unlike the System V printing system, CUPS allows printer names to contain any printable character except SPACE, TAB, "/", or "#". Also, printer and class names are \fInot\fR case-sensitive.
.SH EXAMPLES
Cancel the current print job:
.nf

    cancel

.fi
Cancel job "myprinter-42":
.nf

    cancel myprinter\-42

.fi
Cancel all jobs:
.nf

    cancel \-a
.fi
.SH NOTES
The
.B \-a
and
.BI \-u " username"
options require administrative access.
These restrictions are part of the operation policies defined in the
.BR cupsd.conf (5)
file for the "Cancel-Jobs" and "Cancel-My-Jobs" operations, respectively.
.SH SEE ALSO
.BR cupsd.conf (5),
.BR lp (1),
.BR lpmove (8),
.BR lpstat (1),
CUPS Online Help (http://localhost:631/help)
.SH COPYRIGHT
Copyright \[co] 2020-2024 by OpenPrinting.


CHANGES.md
CHANGES - OpenPrinting CUPS
===========================

v2.5b1 - YYYY-MM-DD
-------------------

- Added multiple language support for IPP Everywhere.
- Added `cupsConcatString`, `cupsCopyString`, and `cupsFormatString` string
  APIs.
- Added `cupsRasterInitHeader` API.
- Added `httpConnectURI`, `httpGetCookieValue`, and `httpGetSecurity` APIs.
- Added `ippAddCredentialsString`, `ippGetFirstAttribute`,
  `ippGetNextAttribute`, `ippRestore`, and `ippSave` APIs.
- Added new DNS-SD APIs.
- Added new JSON APIs.
- Added new JWT APIs.
- Added new OAuth APIs.
- Added new WWW form APIs.
- Added "job-sheets-col" support (Issue #138)
- Added "--list-all" option for cupsfilter (Issue #194)
- Added support for wide-area DNS-SD with Avahi (Issue #319)
- Added `cupsCopyDestInfo2` API (Issue #586)
- Added OpenSSL support for `cupsHashData` (Issue #762)
- Added "job-presets-supported" support for CUPS shared printers (Issue #778)
- Added warning if the device has to do IPP request for 'all,media-col-database'
  in separate requests (Issue #829)
- Added a new argument value for `lpstat` argument '-W' - `successful` -
  to get successfully printed jobs (Issue #830)
- Added driver filter to web interface (Issue #848)
- Added support for PAM modules password-auth and system-auth (Issue #892)
- Added Docker support (Issue #929)
- Added a systemd slice to the systemd services included with the scheduler
- Added localizations for deprecated IPP attributes/options (Issue #1020)
- Added support for specifying permissions with the `cupsFileOpen` API.
- Added `cupsGetClock` API.
- Added `cupsParseOptions2` API with "end" argument.
- Added `cups-oauth` and `cups-x509` utilities (Issue #1184)
- Added `DNSSDComputerName` directive to "cupsd.conf" and updated cupsd to
  correctly update the mDNS hostname only if the `DNSSDHostName` directive is
  not specified (Issue #1217)
- Added `print-as-raster` printer and job attributes for forcing rasterization
  (Issue #1282)
- Added an "install" sub-command to the `cups-x509` command (Issue #1227)
- Added a "--user-agent" option to the `ipptool` command.
- Updated documentation (Issue #984, Issue #1086, Issue #1182)
- Updated translations (Issue #1146, Issue #1161, Issue #1164)
- Updated the configure script to default to installing to /usr/local.
- Updated CUPS to use the Windows mDNS APIs.
- Updated CUPS to require TLS support - OpenSSL, GNUTLS and LibreSSL are
  supported.
- Updated CUPS to require ZLIB.
- Updated CUPS to require support for `poll` API.
- Updated `cupsArray` APIs to support modern naming and types.
- Updated IPP Everywhere printer creation error reporting (Issue #347)
- Updated internal usage of CUPS array API to include callback pointer even when
  not used (Issue #674)
- Updated support for using keyword equivalents for enumerated values like
  "print-quality" (Issue #734)
- Updated `cups_enum_dests()` timeout for listing available IPP printers
  (Issue #751)
- Updated the `ippeveprinter` program to support the `-f` option with `-a`
  (Issue #759)
- Updated default destination documentation (Issue #819)
- Updated `httpAddrConnect2()` to handle `POLLHUP` together with `POLLIN` or
  `POLLOUT` (Issue #839)
- Updated the policies help document with the correct `Order` directive to deny
  access by default (Issue #844)
- Updated the "get-printer-attributes-suite.test" test file (Issue #909)
- Updated `cupsRasterReadPixels` and `cupsRasterWritePixels` to not try reading
  or writing if the number of bytes passed is 0 (Issue #914)
- Updated and documented the MIME typing buffering limit (Issue #925)
- Updated the maximum file descriptor limit for `cupsd` to 64k-1 (Issue #989)
- Updated `httpConnectAgain` to re-validate the server's X.509 certificate
  (Issue #1061)
- Updated the raster functions to report more issues via
  `cupsRasterGetErrorString`.
- Updated the `ipptool` utility to support the `--bearer-token` and
  `--client-name` options.
- Updated `cupsEnumDests` and `cupsGetDests` to support printer browsing and
  filtering options in client.conf (Issue #1180)
- Updated the CUPS web interface to make administrative tasks more discoverable
  (Issue #1207)
- Updated the `httpSetCookie` API to support multiple "Set-Cookie:" header
  values.
- Updated the setuid/gid checks in libcups to use `getauxval` on Linux to avoid
  potential security issues (Issue #1258)
- Updated the ready media support to report the default paper size/source/type
  from the PPD file (Issue #1388)
- Updated the scheduler with a more efficient algorithm for determining the
  supported document formats for a printer (Issue #1392)
- Updated the scheduler to hold password/PIN jobs for 15 seconds before
  releasing to the printer (Issue #1456)
- Updated `httpAddrLookup` to return a numeric address when the resolver
  returns "localhost" for a non-loopback address.
- Deprecated the "page-border" Job Template attribute (Issue #1020)
- Removed the `cups-config` utility (use `pkg-config` instead)
- Fixed use-after-free in `cupsdAcceptClient()` when we log warning during error
  handling (fixes CVE-2023-34241)
- Fixed hanging of `lpstat` on Solaris (Issue #156)
- Fixed mapping of PPD InputSlot, MediaType, and OutputBin values (Issue #238)
- Fixed Digest authentication support (Issue #260)
- Fixed document-unprintable-error support (Issue #391)
- Fixed the web interface not showing an error for a non-existent printer
  (Issue #423)
- Fixed extensive looping in scheduler (Issue #604)
- Fixed printing multiple files on specific printers (Issue #643)
- Fixed printing of jobs with job name longer than 255 chars on older printers
  (Issue #644)
- Fixed segfault in `cupsGetNamedDest()` when trying to get default printer, but
  the default printer is not set (Issue #719)
- Fixed ready media support for iOS 17+ (Issue #738)
- Fixed delays in lpd backend (Issue #741)
- Fixed purging job files via `cancel -x` (Issue #742)
- Fixed RFC 1179 port reserving behavior in LPD backend (Issue #743)

systemv/cancel.c
/*
 * "cancel" command for CUPS.
 *
 * Copyright © 2020-2024 by OpenPrinting.
 * Copyright © 2007-2018 by Apple Inc.
 * Copyright © 1997-2006 by Easy Software Products.
 *
 * Licensed under Apache License v2.0.  See the file "LICENSE" for more
 * information.
 */

/*
 * Include necessary headers...
 */

#include <cups/cups-private.h>


/*
 * Local functions...
 */

static void	usage(void) _CUPS_NORETURN;


/*
 * 'main()' - Parse options and cancel jobs.
 */

int					/* O - Exit status */
main(int  argc,				/* I - Number of command-line arguments */
     char *argv[])			/* I - Command-line arguments */
{
  http_t	*http;			/* HTTP connection to server */
  int		i;			/* Looping var */
  int		job_id;			/* Job ID */
  int		num_dests;		/* Number of destinations */
  cups_dest_t	*dests;			/* Destinations */
  char		*opt,			/* Option pointer */
		*dest,			/* Destination printer */
		*job,			/* Job ID pointer */
		*user;			/* Cancel jobs for a user */
  int		purge;			/* Purge or cancel jobs? */
  char		uri[1024];		/* Printer or job URI */
  ipp_t		*request;		/* IPP request */
  ipp_t		*response;		/* IPP response */
  ipp_op_t	op;			/* Operation */


  _cupsSetLocale(argv);

 /*
  * Setup to cancel individual print jobs...
  */

  op        = IPP_OP_CANCEL_JOB;
  purge     = 0;
  dest      = NULL;
  user      = NULL;
  http      = NULL;
  num_dests = 0;
  dests     = NULL;


 /*
  * Process command-line arguments...
  */

  for (i = 1; i < argc; i ++)
  {
    if (!strcmp(argv[i], "--help"))
      usage();
    else if (argv[i][0] == '-' && argv[i][1])
    {
      for (opt = argv[i] + 1; *opt; opt ++)
      {
	switch (*opt)
	{
	  case 'E' : /* Encrypt */
	      cupsSetEncryption(HTTP_ENCRYPTION_REQUIRED);

	      if (http)
		httpEncryption(http, HTTP_ENCRYPTION_REQUIRED);
	      break;

	  case 'U' : /* Username */
	      if (opt[1] != '\0')
	      {
		cupsSetUser(opt + 1);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;
		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected username after \"-U\" option."), argv[0]);
		  usage();
		}

		cupsSetUser(argv[i]);
	      }
	      break;

	  case 'a' : /* Cancel all jobs */
	      op = purge ? IPP_OP_PURGE_JOBS : IPP_OP_CANCEL_JOBS;
	      break;

	  case 'h' : /* Connect to host */
	      if (http != NULL)
	      {
		httpClose(http);
		http = NULL;
	      }

	      if (opt[1] != '\0')
	      {
		cupsSetServer(opt + 1);
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected hostname after \"-h\" option."), argv[0]);
		  usage();
		}
		else
		  cupsSetServer(argv[i]);
	      }
	      break;

	  case 'u' : /* Username */
	      op = IPP_OP_CANCEL_MY_JOBS;

	      if (opt[1] != '\0')
	      {
		user = opt + 1;
		opt += strlen(opt) - 1;
	      }
	      else
	      {
		i ++;

		if (i >= argc)
		{
		  _cupsLangPrintf(stderr, _("%s: Error - expected username after \"-u\" option."), argv[0]);
		  usage();
		}
		else
		  user = argv[i];
	      }
	      break;

	  case 'x' : /* Purge job(s) */
	      purge = 1;

	      if (op == IPP_OP_CANCEL_JOBS)
		op = IPP_OP_PURGE_JOBS;
	      break;

	  default :
	      _cupsLangPrintf(stderr, _("%s: Error - unknown option \"%c\"."), argv[0], *opt);
	      return (1);
	}
      }
    }
    else
    {
     /*
      * Cancel a job or printer...
      */

      if (num_dests == 0)
        num_dests = cupsGetDests(&dests);

      if (!strcmp(argv[i], "-"))
      {
       /*
        * Delete the current job...
	*/

        dest   = "";
	job_id = 0;
      }
      else if (cupsGetDest(argv[i], NULL, num_dests, dests) != NULL)
      {
       /*
        * Delete the current job on the named destination...
	*/

        dest   = argv[i];
	job_id = 0;
      }
      else if ((job = strrchr(argv[i], '-')) != NULL && isdigit(job[1] & 255))
      {
       /*
        * Delete the specified job ID.
	*/

        dest   = NULL;
	op     = IPP_OP_CANCEL_JOB;
        job_id = atoi(job + 1);
      }
      else if (isdigit(argv[i][0] & 255))
      {
       /*
        * Delete the specified job ID.
	*/

        dest   = NULL;
	op     = IPP_OP_CANCEL_JOB;
        job_id = atoi(argv[i]);
      }
      else
      {
       /*
        * Bad printer name!
	*/

        _cupsLangPrintf(stderr,
	                _("%s: Error - unknown destination \"%s\"."),
			argv[0], argv[i]);
	return (1);
      }

     /*
      * For Solaris LP compatibility, ignore a destination name after
      * cancelling a specific job ID...
      */

      if (job_id && (i + 1) < argc &&
          cupsGetDest(argv[i + 1], NULL, num_dests, dests) != NULL)
        i ++;

     /*
      * Open a connection to the server...
      */

      if (http == NULL)
	if ((http = httpConnectEncrypt(cupsServer(), ippPort(),
	                               cupsEncryption())) == NULL)
	{
	  _cupsLangPrintf(stderr,
	                  _("%s: Unable to connect to server."), argv[0]);
	  return (1);
	}

     /*
      * Build an IPP request, which requires the following
      * attributes:
      *
      *    attributes-charset
      *    attributes-natural-language
      *    printer-uri + job-id *or* job-uri
      *    [requesting-user-name]
      *    [purge-job] or [purge-jobs]
      */

      request = ippNewRequest(op);

      if (dest)
      {
	httpAssembleURIf(HTTP_URI_CODING_ALL, uri, sizeof(uri), "ipp", NULL,
	                 "localhost", 0, "/printers/%s", dest);
	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
	             "printer-uri", NULL, uri);
	ippAddInteger(request, IPP_TAG_OPERATION, IPP_TAG_INTEGER, "job-id",
	              job_id);
      }
      else
      {
        snprintf(uri, sizeof(uri), "ipp://localhost/jobs/%d", job_id);
	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "job-uri", NULL,
	             uri);
      }

      if (user)
      {
	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
                     "requesting-user-name", NULL, user);
	ippAddBoolean(request, IPP_TAG_OPERATION, "my-jobs", 1);

        if (op == IPP_OP_CANCEL_JOBS)
          op = IPP_OP_CANCEL_MY_JOBS;
      }
      else
	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
                     "requesting-user-name", NULL, cupsGetUser());

      if (purge)
      {
	if (op == IPP_OP_CANCEL_JOB)
	  ippAddBoolean(request, IPP_TAG_OPERATION, "purge-job", (char)purge);
	else
	  ippAddBoolean(request, IPP_TAG_OPERATION, "purge-jobs", (char)purge);
      }

     /*
      * Do the request and get back a response...
      */

      if (op == IPP_OP_CANCEL_JOBS && (!user || _cups_strcasecmp(user, cupsGetUser())))
        response = cupsDoRequest(http, request, "/admin/");
      else
        response = cupsDoRequest(http, request, "/jobs/");

      if (cupsGetError() > IPP_STATUS_OK_CONFLICTING)
      {
	_cupsLangPrintf(stderr, _("%s: %s failed: %s"), argv[0],
	        	op == IPP_OP_PURGE_JOBS ? "purge-jobs" : "cancel-job",
        		cupsGetErrorString());

          ippDelete(response);

	return (1);
      }

      ippDelete(response);
    }
  }

  if (num_dests == 0 && op != IPP_OP_CANCEL_JOB)
  {
   /*
    * Open a connection to the server...
    */

    if (http == NULL)
      if ((http = httpConnectEncrypt(cupsServer(), ippPort(),
	                             cupsEncryption())) == NULL)
      {
	_cupsLangPrintf(stderr, _("%s: Unable to contact server."), argv[0]);
	return (1);
      }

   /*
    * Build an IPP request, which requires the following
    * attributes:
    *
    *    attributes-charset
    *    attributes-natural-language
    *    printer-uri + job-id *or* job-uri
    *    [requesting-user-name]
    */

    request = ippNewRequest(op);

    ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
	         "printer-uri", NULL, "ipp://localhost/printers/");

    if (user)
    {
      ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
                   "requesting-user-name", NULL, user);
      ippAddBoolean(request, IPP_TAG_OPERATION, "my-jobs", 1);
    }
    else
      ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
                   "requesting-user-name", NULL, cupsGetUser());

    ippAddBoolean(request, IPP_TAG_OPERATION, "purge-jobs", (char)purge);

   /*
    * Do the request and get back a response...
    */

    response = cupsDoRequest(http, request, "/admin/");

    if (cupsGetError() > IPP_STATUS_OK_CONFLICTING)
    {
      _cupsLangPrintf(stderr, _("%s: %s failed: %s"), argv[0],
		      op == IPP_OP_PURGE_JOBS ? "purge-jobs" : "cancel-job",
        	      cupsGetErrorString());

      ippDelete(response);

      return (1);
    }

    ippDelete(response);
  }

  return (0);
}


/*
 * 'usage()' - Show program usage and exit.
 */

static void
usage(void)
{
  _cupsLangPuts(stdout, _("Usage: cancel [options] [id]\n"
                          "       cancel [options] [destination]\n"
                          "       cancel [options] [destination-id]"));
  _cupsLangPuts(stdout, _("Options:"));
  _cupsLangPuts(stdout, _("-a                      Cancel all jobs"));
  _cupsLangPuts(stdout, _("-E                      Encrypt the connection to the server"));
  _cupsLangPuts(stdout, _("-h server[:port]        Connect to the named server and port"));
  _cupsLangPuts(stdout, _("-u owner                Specify the owner to use for jobs"));
  _cupsLangPuts(stdout, _("-U username             Specify the username to use for authentication"));
  _cupsLangPuts(stdout, _("-x                      Purge jobs rather than just canceling"));

  exit(1);
}
